### PR TITLE
Add predicates to only watch relevant resource changes

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
Some changes in a resource are not relevant for the reoconcile function. Mainly changes in `.status` field.
This change adds a predicate to filter out these changes.

**Which issue(s) this PR fixes**: 
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2082912

**Release note**:
```release-note
None
```

### CODE REVIEW

File: controllers/ssp_controller.go [hunk 1]
```diff
@@ -680,19 +676,43 @@ func watchClusterResources(builder *ctrl.Builder, sspOperands []operands.Operand
...
+	// Deduplicate watches
+	watchedTypes := make(map[reflect.Type]operands.WatchType)
 	for _, operand := range sspOperands {
-		for _, t := range watchTypesFunc(operand) {
-			if _, ok := watchedTypes[reflect.TypeOf(t)]; ok {
-				continue
+		for _, watchType := range watchTypesFunc(operand) {
+			key := reflect.TypeOf(watchType.Object)
+			if wt, ok := watchedTypes[key]; ok {
+				watchType.WatchFullObject = watchType.WatchFullObject || wt.WatchFullObject
```
Comment (0xFelix) [inline controllers/ssp_controller.go:686]: Can you place a comment here, what this is for? I assume it should watch the full object as soon as more than one operand defines a Watch for it?

File: controllers/ssp_controller.go [hunk 2]
```diff
@@ -680,19 +676,43 @@ func watchClusterResources(builder *ctrl.Builder, sspOperands []operands.Operand
...
 
-			builder.Watches(
-				&source.Kind{Type: t},
-				handler_hook.New(handler, hookFunc),
-			)
-			watchedTypes[reflect.TypeOf(t)] = struct{}{}
+	for _, watchType := range watchedTypes {
+		var predicates []predicate.Predicate
+		if !watchType.WatchFullObject {
+			predicates = []predicate.Predicate{relevantChangesPredicate()}
```
Comment (0xFelix) [inline controllers/ssp_controller.go:695]: Maybe the Predicate could be part of WatchType so each operand could precisely define the predicate for its types? No/nil predicate would indicate `WatchFullObject`.

File: internal/operands/common-templates/reconcile.go [hunk 3]
```diff
@@ -35,10 +35,11 @@ func init() {
...
 }
 
-func WatchClusterTypes() []client.Object {
-	return []client.Object{
-		&templatev1.Template{},
-	}
+func WatchClusterTypes() []operands.WatchType {
+	return []operands.WatchType{{
+		Object:          &templatev1.Template{},
+		WatchFullObject: true,
```
Comment (0xFelix) [inline internal/operands/common-templates/reconcile.go:41]: `WatchFullObject = true` is the default?

File: internal/operands/data-sources/reconcile.go [hunk 4]
```diff
@@ -39,14 +39,14 @@ func init() {
...
-	return []client.Object{
-		&rbac.ClusterRole{},
-		&rbac.Role{},
-		&rbac.RoleBinding{},
-		&core.Namespace{},
-		&cdiv1beta1.DataSource{},
-		&cdiv1beta1.DataImportCron{},
+func WatchClusterTypes() []operands.WatchType {
+	return []operands.WatchType{
+		{Object: &rbac.ClusterRole{}, WatchFullObject: true},
```
Comment (0xFelix) [inline internal/operands/data-sources/reconcile.go:44]: `WatchFullObject = true` is the default?

File: internal/operands/data-sources/reconcile.go [hunk 5]
```diff
@@ -39,14 +39,14 @@ func init() {
...
+		{Object: &rbac.Role{}, WatchFullObject: true},
+		{Object: &rbac.RoleBinding{}, WatchFullObject: true},
+		{Object: &core.Namespace{}},
+		{Object: &cdiv1beta1.DataSource{}},
+		{Object: &cdiv1beta1.DataImportCron{}},
```
Comment (0xFelix) [inline internal/operands/data-sources/reconcile.go:49]: `WatchFullObject = true` is the default?

Comment (ksimon1) [reply internal/operands/data-sources/reconcile.go:49]: when bool is not defined, it has always value false

File: controllers/ssp_controller.go (hunk 1 above)
Comment (akrejcir) [reply controllers/ssp_controller.go:686]: Not "more then one operand". If at least one operand wants to watch the full object, then the watch should allow it.

File: controllers/ssp_controller.go (hunk 2 above)
Comment (akrejcir) [reply controllers/ssp_controller.go:695]: I would rather keep it simple. If sometime we will need different predicates for various operands, we can add the code.

File: internal/operands/common-templates/reconcile.go (hunk 3 above)
Comment (akrejcir) [reply internal/operands/common-templates/reconcile.go:41]: No, `false` is default.

File: controllers/ssp_controller.go (hunk 1 above)
Comment (0xFelix) [reply controllers/ssp_controller.go:686]: Ok, a comment would still be helpful.

Comment (akrejcir) [reply controllers/ssp_controller.go:688]: Done

**Comment by akrejcir** [comment]
I need to manually verify that the loop does not happen.
/hold

**Comment by akrejcir** [comment]
This PR alone will not fix the loop, because CDI resources have their `generation` updated also on `status` changes.

/hold cancel

**Comment by akrejcir** [comment]
I will change this, so that it does not assume that `generation` does not update when `status` updates.
/hold

**Comment by akrejcir** [comment]
Holding until I manually verify the loop.
/hold

**Comment by akrejcir** [comment]
The failing tests in CI are related to prometheus configuration in the testing cluster, they are not caused by this PR.

/override ci/prow/e2e-functests
/override ci/prow/e2e-single-node-functests
/override ci/prow/e2e-upgrade-functests

**Comment by akrejcir** [comment]
Karel gave `/lgtm` on the commit with new functionality. The other commit only fixes CI issues unrelated to this PR.
I'm merging this PR.

/override tide

**Pull request merged**